### PR TITLE
Assignment5

### PR DESCRIPTION
**Pull request by FlamingLeo**
Finished worksheet 5.
- Added membrane simulation.
- Added parallelization with OpenMP.
- Added support for 3D linked cell Lennard-Jones simulations.
- Added nanoflow simulations using an updated Thermostat.
- Added out-of-the-box Intel C++ compiler support.
- Added various compiler optimizations.
- Added previous reports and presentations to repository.
- Made Xerces-C mandatory again.

### CODE REVIEW

File: tests/strategies/TestForceCalculation.cpp [hunk 1]
```diff
@@ -51,4 +51,110 @@ TEST_F(ForceTests, UpdateForceLJ) {
...
+                      1,
+                      pc2,
+                      2};
+    calculateF_LennardJones_LC(pc1, 0, &cc1);
+    calculateF_LennardJones_LC_task(pc2, 0, &cc2);
+    ASSERT_EQ(pc1.size(), pc2.size());
+    for (size_t i = 0; i < pc1.size(); ++i) {
+        EXPECT_EQ(pc1[i], pc2[i]);
+    }
+}
```
Comment (SamNewcome) [inline tests/strategies/TestForceCalculation.cpp]: When I run this, none of the particles with either method have any force - there are no force calculations so the parallelisation is not really tested

File: src/objects/Cell.cpp [hunk 2]
```diff
@@ -8,55 +8,164 @@
...
+    RETURN_TRIPLE_DIAG_REVERSE(BNW, ((1 - c) * x * relX + (1 - c) * x * relZ + c * y * relY >= c * y * y),
+                               (relX <= relZ), HaloLocation::NORTH, HaloLocation::WEST, HaloLocation::BELOW);
+    RETURN_TRIPLE_DIAG_CONDITION(ASE, ((1 - c) * x * relX + (1 - c) * x * relZ + c * y * relY >= c * y * y),
+                                 (relX <= relZ), HaloLocation::ABOVE, HaloLocation::EAST, HaloLocation::SOUTH);
+    RETURN_TRIPLE_DIAG_CONDITION(ASW, ((c - 1) * x * relX + (1 - c) * z * relZ + c * y * relY >= (1 - c) * z * z),
+                                 (relX > relZ), HaloLocation::ABOVE, HaloLocation::WEST, HaloLocation::SOUTH);
+    RETURN_TRIPLE_DIAG_REVERSE(ANE, ((c - 1) * x * relX + (c - 1) * z * relZ + c * y * relY >= 0), (relX <= relZ),
+                               HaloLocation::NORTH, HaloLocation::ABOVE, HaloLocation::EAST);
+    RETURN_TRIPLE_DIAG_REVERSE(ANW, ((1 * c) * x * relX + (c - 1) * z * relZ + c * y * relY >= (1 - c) * x * x),
+                               (relX > relZ), HaloLocation::NORTH, HaloLocation::ABOVE, HaloLocation::WEST);
```
Comment (SamNewcome) [inline src/objects/Cell.cpp:144]: This logic is really hard to follow. There is no description of what  these do and the names don't make this clear enough.

Also, I don't really like the macro-based solution here, and they are generally considered bad practice unless there's a good reason (see https://www.fluentcpp.com/2019/05/14/3-types-of-macros-that-improve-c-code/) 

The nicer solution, in my opinion, would be to use helper functions. These can be inlined if you are worried about the cost of calling the function.

File: src/objects/Cell.cpp [hunk 3]
```diff
@@ -8,55 +8,164 @@
...
+    bool ASE = isCorner(HaloLocation::ABOVE, HaloLocation::SOUTH) && VEC_CONTAINS(m_haloLocation, HaloLocation::EAST);
+    bool ASW = isCorner(HaloLocation::ABOVE, HaloLocation::SOUTH) && VEC_CONTAINS(m_haloLocation, HaloLocation::WEST);
+    bool ANE = isCorner(HaloLocation::ABOVE, HaloLocation::NORTH) && VEC_CONTAINS(m_haloLocation, HaloLocation::EAST);
+    bool ANW = isCorner(HaloLocation::ABOVE, HaloLocation::NORTH) && VEC_CONTAINS(m_haloLocation, HaloLocation::WEST);
+    double relX = relPos[0], relY = relPos[1], relZ = relPos[2];
+
+    double x = m_size[0];
+    double y = m_size[1];
+    double z = m_size[2];
+    double c = (m_size[0] * m_size[0] + m_size[2] * m_size[2]) /
```
Comment (SamNewcome) [inline src/objects/Cell.cpp:126]: `c` is not a descriptive enough name. Nor is it really clearer what it's for.

File: tests/strategies/TestBoundaryConditions.cpp [hunk 4]
```diff
@@ -205,4 +208,221 @@ TEST(BoundaryConditionTests, MixedReflectivePeriodicCorner) {
...
     test({1.25, 1.25, 0.}, {-20., -20., 0.}, {5.25, 1.75, 0.}, {-20., 20., 0.}, 12); // sw corner
     test({1.25, 5.75, 0.}, {-20., 20., 0.}, {5.25, 5.25, 0.}, {-20., -20., 0.}, 40); // nw corner
-}
\ No newline at end of file
+}
+
+// CONCERNING 3D TESTS:
+// 3D tests do not cover all boundaries and boundary types anymore as it's not in my opinion feasible
+// (there's for example 20 distinct corner cases that would need individual testing, which seems a bit much)
+// Rather, tests now only sample from the main categories (double corner etc.)
```
Comment (SamNewcome) [inline tests/strategies/TestBoundaryConditions.cpp:216]: For the purposes of this course, I'm happy with these tests.

But I don't think that this is infeasible. I totally get that making "hand-calculated" values for each case is infeasible, but you can also just test against some otherwise computed solution. You know the effect of each individual BC, so you can calculate what the effect of all of them combined should be. You can then handle all the cases using gtest's parameterised tests or just with a simple lambda function, which allows you to easily make sure all combinations of particle position/behaviour and boundary types are covered.

File: src/objects/Particle.h [hunk 5]
```diff
@@ -10,12 +10,37 @@
...
+    if ((_p).getType() == 1)                                                                                           \
+        continue;
+
+#ifdef NOUTFLOW
+#define CONTINUE_IF_INACTIVE(p) (void)0
+#else
+#define CONTINUE_IF_INACTIVE(p)                                                                                        \
+    if (!(p).isActive())                                                                                               \
+        continue;
+#endif
```
Comment (SamNewcome) [inline src/objects/Particle.h]: Using macros are often dangerous. There are two key reasons: (1) they don't do any type checking, and (2) if included in a header file, the definitions can proliferate throughout the code, similar to why we suggest no global level variables. Sometimes, there are times when they are the best solution (even if problematic) - particularly if the code is dependent on some CMake-defined macro. But you should use the following alternatives where possible:

- If you have fixed values, e.g. like between lines 18 and 24, you can use `constexpr` variable definitions that do the same things. These can be put inside the class to avoid the definitions proliferating throughout the code. This also allows you to give the variable a type which makes it much easier to check that this type is used correctly.
- If you don't know the type at time of writing e.g. if you don't know that you will use a `double` for epsilon, you can make `Particle` a class templated with a `epsilonType` which can be specified to `double` at compile time when `Particle` is instantiated. Similarly, for functions where you don't know the type of some variable at the time of writing or if you intend to use them with multiple different types. 
- Here you are really just using this to skip the activity check - you could just include `#ifndef NOUTFLOW`to wrap this activity check within the code itself. This does the same thing but avoid using globally scoped macro functions.

File: CMakeLists.txt [hunk 6]
```diff
@@ -28,11 +28,20 @@ endif()
...
 include(gtest)
 include(benchmark)
 include(spdlog)
+include(pgo)
+
+# add (potentially dangerous) option to ensure that there are no outflow conditions
+option(NO_OUTFLOW "Ensure no outflow simulations" OFF)
+if(NO_OUTFLOW)
+    message(STATUS "Outflow simulations DISABLED! Be careful!")
+    add_compile_definitions(NOUTFLOW)
```
Comment (SamNewcome) [inline CMakeLists.txt:43]: `NOUTFLOW` is not a super clear macro name. `NO_OUTFLOW` is much clearer.

File: src/objects/Thermostat.cpp [hunk 7]
```diff
@@ -1,64 +1,77 @@
...
+            CONTINUE_IF_INACTIVE(p);
+            DO_IF_NOT_WALL(p, sum += p.getM() * ArrayUtils::L2NormSquared(p.getThermalMotion()));
+        }
     }
     kineticEnergy = sum / 2;
     SPDLOG_TRACE("New kinetic energy: {}", kineticEnergy);
 }
 void Thermostat::calculateTemp() {
-    temperature = 2 * kineticEnergy / (dimension * particles.size());
+    temperature = 2 * kineticEnergy / (dimension * particles.nonWallSize());
```
Comment (SamNewcome) [inline src/objects/Thermostat.cpp:74]: There is an implicit narrowing conversion here to double. From personal experience, these can lead to unexpected problems and painful debugging as they can be handled differently by different compilers. Use something like `static_cast` to really ensure that the code converts types the way you intend it to.

File: src/objects/CellContainer.h [hunk 8]
```diff
@@ -391,6 +419,14 @@ class CellContainer {
      */
     std::vector<int> getOppositeOfBorderCorner(const Cell &from, std::vector<BorderLocation> &locations);
 
+    /**
+     * @brief For a collection of border locations, returns all unique pairs
+     *
+     * @param locations The border locations/directions which form the pairs
+     * @return A vector of vectors representing the unique pairs
```
Comment (SamNewcome) [inline src/objects/CellContainer.h]: It is not clear what "pairs" are

File: src/objects/Particle.h (hunk 5 above)
Comment (SamNewcome) [reply src/objects/Particle.h]: Furthermore, there should still be documentation to explain what they do, even if it does not have to be standard doxygen function documentation.

File: src/objects/Thermostat.cpp (hunk 7 above)
Comment (SamNewcome) [reply src/objects/Thermostat.cpp:74]: This is not the only place in your code with this issue.

File: src/strategies/BoundaryConditions.cpp [hunk 9]
```diff
@@ -135,19 +136,21 @@ void mirrorGhostParticles(CellContainer *lc) {
...
         std::vector<BorderLocation> location = bc.get().getBorderLocation();
-        // special case for corners (2D) -- should be changed to >2 for 3D
+
+        // I'm starting to think it should be 1 even for 3D (imagine a cross section)
         if (location.size() > 1) {
-            // if not all concerned edges are periodic, don't mirror corner (intuitively this is how it seems it should
-            // be to me)
-            bool doCorner = true;
+            // At least 2 edges should be periodic for us to mirror in corner (intuition)
+            int doCorner = 0;
```
Comment (SamNewcome) [inline src/strategies/BoundaryConditions.cpp:143]: Having `doCorner` as an `int` doesn't really make sense. Really, this is the number of periodic boundaries along this cell, right?

**Review by SamNewcome — APPROVED** [review]
Well done! The final code is looking really nice.

My only significant issue is with the heavy use of macros where they are generally inappropriate. This is not terrible but bad practice for the reasons described in the code. The exception is your container for-looping macros - although I still find these questionable if not totally inappropriate.

File: src/objects/CellContainer.h (hunk 8 above)
Comment (FlamingLeo) [reply src/objects/CellContainer.h]: The combinations of size 2 using the cardinal directions passed to the function as input.
I've updated the description to better represent this.

File: src/strategies/BoundaryConditions.cpp (hunk 9 above)
Comment (FlamingLeo) [reply src/strategies/BoundaryConditions.cpp:143]: IIRC this was more intuition-based than anything, and we counted the number of periodic boundaries to determine whether we even needed to mirror ghost particles in corner cells (potentially for edge cases where only one side is periodic, but the opposite isn't?). In hindsight, the check seems rudimentary at best and pointless at worst, but the code works so I'm afraid to touch it...

File: CMakeLists.txt (hunk 6 above)
Comment (FlamingLeo) [reply CMakeLists.txt:43]: I agree; I instinctively named it `NOUTFLOW` based on `NDEBUG` (Not Debug).

File: src/objects/Cell.cpp (hunk 2 above)
Comment (FlamingLeo) [reply src/objects/Cell.cpp:144]: If the logic seems hard to follow, that's because it is (which is a crap answer, I know, but we kinda dug our own graves when we decided to go this route for doing boundary conditions). Essentially, it's the 2D algorithm for determining which boundary a particle hits first, adjusted for 3D using two planes similar to the delimiting line used for 2D simulations, and checking the relative coordinates inside the cell to see where the particle lands in relation to the two halving planes (@MaraGodeanu I think the raw details of the geometrical calculations used were skipped in the report?).

As for the macros, these were mostly a last-minute fix to remove _heavy_ repetition. I agree that they should be swapped for functions, but since most of the code was just copy-pasted with swapped conditions (and especially since the function returns inside these conditions), I just blindly chose to use macros.

File: src/objects/Cell.cpp (hunk 3 above)
Comment (FlamingLeo) [reply src/objects/Cell.cpp:126]: @MaraGodeanu ?

File: src/objects/Particle.h (hunk 5 above)
Comment (FlamingLeo) [reply src/objects/Particle.h]: To be fair, the usage of macros mainly comes from my prior experience with C, where macros are still pretty heavily used in some codebases. But in most cases, they should be avoided in C++, agreed.

- `constexpr` could have been a viable alternative, but I wanted to avoid having to write a bunch of getter functions including unique documentation and probabily having to instantiate / rely on instantiated objects just to get a common variable (though I guess `static` definitions could have worked?).
- We did use `#ifndef NOUTFLOW` throughout the code at first, but that was in my opinion a lot of needless repetition. A macro just makes things cleaner here, albeit at the (minimal) risk of using a global name.

I've also added documentation since.

File: src/objects/Thermostat.cpp (hunk 7 above)
Comment (FlamingLeo) [reply src/objects/Thermostat.cpp:74]: Where exactly would this lead to _narrowing_ conversion? `temperature` and `kineticEnergy` are both `double`s, so that's fine; `dimension` is an `int` and `particles.size()` is a `size_t`, but according to the C++ standard on "usual arithmetic conversions", the `int` gets promoted to a `size_t`, which is a larger datatype. Then, this gets converted to a `double` for the final division. As far as I can tell, nowhere along these lines does any value get converted to a type which can't hold it, barring the obvious floating point imprecisions for larger values, which shouldn't happen here anyway since temperatures are typically in the tens or hundreds at most here.

**Comment by FlamingLeo** [comment]
I'm glad you're (mostly) happy with our code, it was a lot of fun working on this project :)

Sorry from my part for not coming to the gathering yesterday - I was (and still am) pretty sick with the flu.

File: src/objects/Thermostat.cpp (hunk 7 above)
Comment (SamNewcome) [reply src/objects/Thermostat.cpp:74]: So you are correct that for the ranges of values that you are using, there shouldn't be any "narrowing". The issue is more with the implicit-ness as opposed to the conversion. Generally, it is considered good practice to use e.g. a `static_cast<double>` to _explicitly_ instruct the compiler to do so.  Here you are effectively telling the compiler that you know what you are doing and so the compiler/clang-tidy will not give a warning. It is important that the compiler gives these warnings for implicit conversions because, in other cases, lack of care with type conversions can lead to problems.

**Comment by SamNewcome** [comment]
> I'm glad you're (mostly) happy with our code, it was a lot of fun working on this project :)
> 
> Sorry from my part for not coming to the gathering yesterday - I was (and still am) pretty sick with the flu.

I'm glad that you liked the project! Don't worry about not coming to the gathering - get well soon!

File: src/objects/Cell.cpp (hunk 3 above)
Comment (MaraGodeanu) [reply src/objects/Cell.cpp:126]: Hi, I'm a bit late to the party, but I'll try to leave some explanations in general for the corners in case you are still interested. I'll preface by saying I gave the boundary decision system for triple corners (the 8 "true" corners of a 3D rectangular parallelepiped) a big-ish overhaul before the deadline so the form of the code leaves something to be desired. I'll be referring to Figures 9 and 10 in the report for a visual aid and be using the corner there as an examples.

So, very late I realized the equations I had written for the green plane only worked in a very specific case, when the cells are a 1x1x1 cube, and thus sought a more general formula for all cell sizes. A plane can be defined by a point on the plane and the plane normal. As shown in Figure 10, I wanted to define the plane so that the line AB _lies in_ the plane, and the perpendicular line from C to AB is the plane normal. 

In the following, I will consider the lower left corner closer to the viewer to have position (0, 0, 0) as positions inside the cell are relative. Consequently, A has position (0, m_size[1], m_size[2]), B has position (m_size[0], 0, 0) and C is (m_size[0], m_size[1], 0). m_size are the cell dimensions and they were renamed in the code to x, y, z for brevity.

Now comes the math I chose _not_ to include in the report because it's tedious:

Vector AB = D = B - A = (x, -y, -z)

Helper Variable V = C - A = (x, 0, -z)

Projection P = ((V dot D) / (D dot D)) * D = ((x^2 + z^2) / (x^2 + y^2 + z^2)) * (x, -y, -z)

...it turns out ((V dot D) / (D dot D)) = ((x^2 + z^2) / (x^2 + y^2 + z^2)) holds for all 4 corner situations so I - rather unoriginally - renamed this value **c** for constant.

Normal N = V - P = (x - c * x, 0 +c * x, -z + c * x) = ((1 - c) * x, c *y, (c - 1) * z)

With this, most work is complete. In order to find out the plane equation one only needs to dot product this with the value of a point _on_ the plane (e.g. B): B dot N = (1 - c) * x * x

From this follows the plane equation: P dot N = (1 - c) * x * x, where P is any point (x_P, y_P, z_P). 

If P dot N >= (1 - c) * x * x then P is on the side of the normal and thus above the plane. 

If you don't mind, I won't describe the 2-plane system again, since it's in the report. In case you wish to have the calculations for all corners, I have them handwritten on my tablet and I can send you the file. 

All in all, as you can see, this was very mathematical and quite a pain to read, so I didn't want to include it in the report. On a personal note, it was also painful for me to make this system, as I have not done any geometry since high school. If I ever do simulation projects again, I will definitely take the iterating through the boundaries approach you showed last week.

I'll echo Flavius' sentiment and say that this was a great Praktikum!

Comment (SamNewcome) [reply src/objects/Cell.cpp:126]: Thanks for the explanation! I'm glad that you enjoyed the course!
